rust/src/domains/polars/frame/ffi.rs

```diff
@@ -0,0 +1,202 @@
+fn infer_lazyframe_domain(lazyframe: LazyFrame) -> Fallible<LazyFrameDomain> {
+    LazyFrameDomain::new_from_schema(lazyframe.schema()?.as_ref().clone())
+}
+
+#[no_mangle]
+pub extern "C" fn opendp_domains__infer_lazyframe_domain(
+    lazyframe: *mut AnyObject,
+) -> FfiResult<*mut AnyDomain> {
+    let lazyframe = try_!(try_as_ref!(lazyframe).downcast_ref::<LazyFrame>()).clone();
+    Ok(AnyDomain::new(try_!(infer_lazyframe_domain(lazyframe)))).into()
+}
+
+fn unpack_series_domains(series_domains: *mut AnyObject) -> Fallible<Vec<SeriesDomain>> {
+    let vec_any = try_as_ref!(series_domains).downcast_ref::<Vec<AnyDomainPtr>>()?;
+
+    vec_any
+        .iter()
+        .map(|x| {
+            util::as_ref(x.clone())
+                .and_then(|ad| ad.downcast_ref::<SeriesDomain>().ok())
+                .cloned()
+        })
+        .collect::<Option<Vec<SeriesDomain>>>()
+        .ok_or_else(|| err!(FailedCast, "domain downcast failed"))
+}
+
+#[bootstrap(
+    name = "with_margin",
+    arguments(
+        frame_domain(rust_type = b"null"),
+        by(rust_type = "Vec<String>"),
+        max_partition_size(rust_type = "Option<u32>", default = b"null"),
+        max_num_partitions(rust_type = "Option<u32>", default = b"null"),
+        max_partition_contributions(rust_type = "Option<u32>", default = b"null"),
+        max_influenced_partitions(rust_type = "Option<u32>", default = b"null"),
+        public_info(rust_type = "Option<String>", default = b"null")
+    ),
+    returns(c_type = "FfiResult<AnyDomain *>")
+)]
+#[no_mangle]
+pub extern "C" fn opendp_domains__with_margin(
+    frame_domain: *mut AnyDomain,
+    by: *mut AnyObject,
+    max_partition_size: *mut AnyObject,
+    max_num_partitions: *mut AnyObject,
+    max_partition_contributions: *mut AnyObject,
+    max_influenced_partitions: *mut AnyObject,
+    public_info: *mut c_char,
+) -> FfiResult<*mut AnyDomain> {
+    let domain = try_as_ref!(frame_domain);
+    let by = try_!(try_as_ref!(by).downcast_ref::<Vec<String>>()).clone();
+
+    let max_partition_size = if let Some(x) = util::as_ref(max_partition_size) {
+        Some(*try_!(x.downcast_ref::<u32>()))
+    } else {
+        None
+    };
+    let max_num_partitions = if let Some(x) = util::as_ref(max_num_partitions) {
+        Some(*try_!(x.downcast_ref::<u32>()))
+    } else {
+        None
+    };
+
+    let max_partition_contributions = if let Some(x) = util::as_ref(max_partition_contributions) {
+        Some(*try_!(x.downcast_ref::<u32>()))
+    } else {
+        None
+    };
+    let max_influenced_partitions = if let Some(x) = util::as_ref(max_influenced_partitions) {
+        Some(*try_!(x.downcast_ref::<u32>()))
+    } else {
+        None
+    };
+
+    let public_info = if let Some(public_info) = try_!(to_option_str(public_info)) {
+        Some(match public_info {
+            "keys" => MarginPub::Keys,
+            "lengths" => MarginPub::Lengths,
```

Comment:
This corresponds to this test helper?
```
def _with_margin(domain, by):
    return dp.with_margin(domain, by=by, public_info="lengths", max_partition_size=50)
```

I feel weird about passing in strings when we really just need to convey a boolean.

Also: It looks like "lengths" is exercised in the tests. Should we have some coverage for "keys"?

Reply:
There's actually three states: no public info, known keys, and known lengths. I think the stringly-typed argument is clearer. An enum would be best, but I don't have FFI for it.

Reply:
Added another test for initializing with keys.